src/query/service/src/pipelines/pipeline_builder.rs

```diff
@@ -708,6 +711,85 @@ impl PipelineBuilder {
         Ok(params)
     }
 
+    fn build_window(&mut self, window: &Window) -> Result<()> {
+        self.build_pipeline(&window.input)?;
+
+        let input_schema = window.input.output_schema()?;
+
+        if !window.order_by.is_empty() {
+            let sort_desc = window
+                .order_by
+                .iter()
+                .map(|desc| SortColumnDescription {
+                    offset: desc.order_by,
+                    asc: desc.asc,
+                    nulls_first: desc.nulls_first,
+                })
+                .collect::<Vec<_>>();
+
+            let max_threads = self.ctx.get_settings().get_max_threads()? as usize;
+            let block_size = self.ctx.get_settings().get_max_block_size()? as usize;
+
+            // TODO(Winter): the query will hang in MultiSortMergeProcessor when max_threads == 1 and output_len != 1
+            if self.main_pipeline.output_len() == 1 || max_threads == 1 {
+                self.main_pipeline.resize(max_threads)?;
+            }
+
+            // Sort
+            self.main_pipeline.add_transform(|input, output| {
+                let transform =
+                    TransformSortPartial::try_create(input, output, None, sort_desc.clone())?;
+
+                Ok(ProcessorPtr::create(transform))
+            })?;
+
+            // Merge
+            self.main_pipeline.add_transform(|input, output| {
+                let transform = try_create_transform_sort_merge(
+                    input,
+                    output,
+                    input_schema.clone(),
+                    block_size,
+                    None,
+                    sort_desc.clone(),
+                )?;
+
+                Ok(ProcessorPtr::create(transform))
+            })?;
+
+            // Concat merge in single thread
+            try_add_multi_sort_merge(
+                &mut self.main_pipeline,
+                input_schema,
+                block_size,
+                None,
+                sort_desc,
+            )?;
+        }
+
+        // let input_schema = window.input.output_schema()?;
+        let agg_func = AggregateFunctionFactory::instance().get(
+            window.agg_func.sig.name.as_str(),
+            window.agg_func.sig.params.clone(),
+            window.agg_func.sig.args.clone(),
+        )?;
+
+        let arguments = window.agg_func.args.clone();
+
+        // Window
+        self.main_pipeline.add_transform(|input, output| {
+            let transform = TransformWindow::try_create(
+                input,
+                output,
+                agg_func.clone(),
+                arguments.clone(),
+                window.partition_by.clone(),
```

Comment:
The column indices in `window.partition_by` should be replaced by the offsets in the input schema. So does `window.order_by`.